SourceX/DiabloUI/diabloui.cpp

```diff
@@ -577,27 +599,27 @@ BOOL UiCreatePlayerDescription(_uiheroinfo *info, DWORD mode, char *desc)
 
 void DrawArt(int screenX, int screenY, Art *art, int nFrame, DWORD drawW)
 {
-	if (screenY >= SCREEN_WIDTH || screenX >= SCREEN_WIDTH)
+	if (screenY >= SCREEN_Y + SCREEN_HEIGHT || screenX >= SCREEN_X + SCREEN_WIDTH)
 		return;
 
-	BYTE *__restrict src = (BYTE *)art->data + (art->width * art->height * nFrame);
-	BYTE *__restrict dst = &gpBuffer[screenX + 64 + (screenY + SCREEN_Y) * BUFFER_WIDTH];
-	drawW = drawW ? drawW : art->width;
+	SDL_Rect src_rect = { 0, nFrame * art->h(), art->w(), art->h() };
+	if (drawW && drawW < src_rect.w)
+		src_rect.w = drawW;
+	SDL_Rect dst_rect = { screenX + SCREEN_X, screenY + SCREEN_Y, src_rect.w, src_rect.h };
 
-	const DWORD iSize = std::min(art->height, static_cast<DWORD>(SCREEN_HEIGHT - screenY));
-	const DWORD jSize = std::min(art->width, std::min(drawW, static_cast<DWORD>(SCREEN_WIDTH - screenX)));
+	if (art->surface->format->BitsPerPixel == 8 && art->palette_version != pal_surface_palette_version) {
+#ifdef USE_SDL1
+		if (SDL_SetPalette(art->surface, SDL_LOGPAL, pal_surface->format->palette->colors, 0, 256) != 1)
+			SDL_Log(SDL_GetError());
+#else
+		if (SDL_SetSurfacePalette(art->surface, pal_surface->format->palette) <= -1)
+			SDL_Log(SDL_GetError());
+#endif
+		art->palette_version = pal_surface_palette_version;
+	}
 
-	if (art->masked) {
-		for (DWORD i = 0; i < iSize; i++, src += art->width, dst += BUFFER_WIDTH) {
-			for (DWORD j = 0; j < jSize; j++) {
-				if (src[j] != art->mask)
-					dst[j] = src[j];
-			}
-		}
-	} else {
-		for (DWORD i = 0; i < iSize; i++, src += art->width, dst += BUFFER_WIDTH) {
-			memcpy(dst, src, jSize * sizeof(dst[0]));
-		}
+	if (SDL_BlitSurface(art->surface, &src_rect, pal_surface, &dst_rect) <= -1) {

```

Comment:
`SDL_RLEACCEL` probably doesn't make sense before we switch to blitting to `surface` instead of `pal_surface` and probably also convert the surfaces to the native display format on load.

Reply:
I think it might not make sense even then: these surfaces are tiny and the branches in RLE decompression are likely slow than the straightforward approach.

Reply:
You might be right. Ironically the Diablo renderer works in a similar fashion, it is hugely over-complicated in an attempt to micro optimize (in ASM no less) our test with rewriting parts of it has shown that a simple generic C render is faster in many cases, even when using the original compiler. They also have all sprites loaded with RLE, but that might have made sense back then since it would save on memory.